File: dotnet/src/Mekik.Core/Scaling.cs
namespace Mekik;

// The horizontal-scale ports (docs/SCALING.md), mirror of the TypeScript
// `scaling.ts`. Like the persistence ports in `Stores.cs`, each is an interface
// with an in-memory default, so a single-node run needs nothing and a fleet swaps
// implementations in. Passing neither keeps mekik's process-local behaviour
// byte-for-byte — scaling is entirely opt-in.
//
// The shapes are native .NET: a lease and a subscription are `IAsyncDisposable`
// (release / unsubscribe on `await using` / `DisposeAsync`), acquisition is
// nullable-returning, and everything is cancellable.

/// <summary>
/// A held per-conversation turn lease. Renewed while a long run streams; released
/// on <see cref="IAsyncDisposable.DisposeAsync"/> so the next turn can proceed.
/// </summary>
public interface ITurnLease : IAsyncDisposable
{
    /// <summary>Extend the lease TTL (a Redis lock heartbeat; a no-op for the local lock).</summary>
    Task RenewAsync(CancellationToken cancellationToken = default);
}

/// <summary>
/// The distributed turn lock. <see cref="AcquireAsync"/> returns a lease, or
/// <c>null</c> when another node already owns the turn for this conversation (the
/// caller answers <c>busy</c>).
/// </summary>
public interface ITurnLock
{
    Task<ITurnLease?> AcquireAsync(string conversationId, CancellationToken cancellationToken = default);
}

/// <summary>
/// The single-node default: always grants. The engine's local turn lock is the
/// real lock on one node, so this lease is a no-op. A Redis <c>SET NX PX</c> lock
/// is the fleet implementation (docs/SCALING.md §The ports).
/// </summary>
public sealed class LocalTurnLock : ITurnLock
{
    private sealed class NoopLease : ITurnLease
    {
        public Task RenewAsync(CancellationToken cancellationToken = default) => Task.CompletedTask;
        public ValueTask DisposeAsync() => ValueTask.CompletedTask;
    }

    public Task<ITurnLease?> AcquireAsync(string conversationId, CancellationToken cancellationToken = default) =>
        Task.FromResult<ITurnLease?>(new NoopLease());
}

/// <summary>
/// What travels across the backplane: a frame plus the node that produced it, so a
/// subscriber can skip its own messages (pub/sub self-delivery guard).
/// </summary>
public sealed record BackplaneMessage(string OriginId, IReadOnlyDictionary<string, object?> Frame);

/// <summary>
/// Cross-node fan-out. The engine <see cref="PublishAsync"/>es every dispatched
/// frame; each node holding a tab of that conversation <see cref="SubscribeAsync"/>s
/// and re-fans the frame to its own sockets. Persist-once stays with the producing
/// node — the backplane only moves already-recorded frames. The returned
/// <see cref="IAsyncDisposable"/> unsubscribes.
/// </summary>
public interface IBackplane
{
    Task PublishAsync(string conversationId, BackplaneMessage message, CancellationToken cancellationToken = default);
    Task<IAsyncDisposable> SubscribeAsync(string conversationId, Action<BackplaneMessage> handler, CancellationToken cancellationToken = default);
}

/// <summary>
/// The single-node default: nothing to carry, because one node's connections already
/// hold every tab. <see cref="PublishAsync"/> drops the message and
/// <see cref="SubscribeAsync"/> never delivers, so behaviour is identical to
/// pre-scaling mekik. Redis Pub/Sub is the fleet implementation.
/// </summary>
public sealed class NoopBackplane : IBackplane
{
    private sealed class NoopSubscription : IAsyncDisposable
    {
        public ValueTask DisposeAsync() => ValueTask.CompletedTask;
    }

    public Task PublishAsync(string conversationId, BackplaneMessage message, CancellationToken cancellationToken = default) =>
        Task.CompletedTask;

    public Task<IAsyncDisposable> SubscribeAsync(string conversationId, Action<BackplaneMessage> handler, CancellationToken cancellationToken = default) =>
        Task.FromResult<IAsyncDisposable>(new NoopSubscription());
}

File: ts/packages/core/src/scaling.ts
// The horizontal-scale ports (docs/SCALING.md). Like the persistence ports in
// `stores.ts`, each is an interface with an in-memory default, so a single-node
// run needs nothing and a fleet swaps implementations in. Passing neither option
// keeps mekik's original process-local behaviour byte-for-byte — scaling is
// entirely opt-in.
//
// Two concerns become fleet-aware:
//   • TurnLock  - the per-conversation single-writer lease. On one node the
//     engine's `live.turn` already guarantees a single writer; a Redis lease
//     extends that guarantee across nodes.
//   • Backplane - cross-node fan-out. On one node the engine fans out directly
//     from `live.connections`; a Redis Pub/Sub backplane carries frames to the
//     other nodes holding tabs of the same conversation.

import type { OutgoingFrame } from "./protocol.ts";

// ── turn lock ─────────────────────────────────────────────────────────────────

/** A held per-conversation turn lease. Renewed while a long run streams; released when it ends. */
export interface TurnLease {
    /** Extend the lease TTL (a Redis lock heartbeat; a no-op for the local lock). */
    renew(): Promise<void>;
    /** Give up the lease so the next turn — here or on another node — can proceed. */
    release(): Promise<void>;
}

/**
 * The distributed turn lock. `acquire` returns a lease, or `null` if another node
 * already holds the turn for this conversation (the caller answers `busy`).
 */
export interface TurnLock {
    acquire(conversationId: string): Promise<TurnLease | null>;
}

/**
 * The single-node default: always grants. The engine's local `live.turn` is the
 * real lock on one node, so this lease is a no-op. A Redis `SET NX PX` lock is the
 * fleet implementation (docs/SCALING.md §The ports).
 */
export class LocalTurnLock implements TurnLock {
    async acquire(_conversationId: string): Promise<TurnLease> {
        return { renew: async () => {}, release: async () => {} };
    }
}

// ── backplane ─────────────────────────────────────────────────────────────────

/** What travels across the backplane: a frame plus the node that produced it (self-delivery guard). */
export interface BackplaneMessage {
    /** The producing node's id — a subscriber skips its own messages. */
    originId: string;
    frame: OutgoingFrame;
}

/** A live backplane subscription for one conversation. */
export interface Subscription {
    unsubscribe(): Promise<void>;
}

/**
 * Cross-node fan-out. The engine `publish`es every dispatched frame; each node
 * holding a tab of that conversation `subscribe`s and re-fans the frame to its
 * own sockets. Persist-once stays with the producing node — the backplane only
 * moves already-recorded frames.
 */
export interface Backplane {
    publish(conversationId: string, message: BackplaneMessage): Promise<void>;
    subscribe(conversationId: string, handler: (message: BackplaneMessage) => void): Promise<Subscription>;
}

/**
 * The single-node default: nothing to carry, because one node's `live.connections`
 * already holds every tab. `publish` drops the message and `subscribe` never
 * delivers, so behaviour is identical to pre-scaling mekik. Redis Pub/Sub is the
 * fleet implementation.
 */
export class NoopBackplane implements Backplane {
    async publish(_conversationId: string, _message: BackplaneMessage): Promise<void> {}
    async subscribe(_conversationId: string, _handler: (message: BackplaneMessage) => void): Promise<Subscription> {
        return { unsubscribe: async () => {} };
    }
}
